### PR TITLE
executor: support projection as inner child of index join

### PR DESCRIPTION
**Pull request by lzmhhh123**
### What problem does this PR solve? 
Fix bug that the master branch can't execute the plan of TPCH Q10.

### What is changed and how it works?
support projection as the inner child of index join

### Check List 

Tests 

 - Unit test
 - Integration test

Side effects

 - Possible performance regression
 - Increased code complexity

### CODE REVIEW

File: executor/builder.go [hunk 1]
```diff
@@ -2121,6 +2123,35 @@ func (builder *dataReaderBuilder) buildIndexLookUpReaderForIndexJoin(ctx context
 	return e, err
 }
 
+func (builder *dataReaderBuilder) buildProjectionForIndexJoin(ctx context.Context, v *plannercore.PhysicalProjection,
+	lookUpContents []*indexJoinLookUpContent, indexRanges []*ranger.Range, keyOff2IdxOff []int, cwc *plannercore.ColWithCmpFuncManager) (Executor, error) {
+	physicalIndexLookUp, isDoubleRead := v.Children()[0].(*plannercore.PhysicalIndexLookUpReader)
+	if !isDoubleRead {
+		return nil, errors.Errorf("it's a invalid plan when the projection is the inner child of indexMergeJoin and its child is not indexLookUpReader")
```
Comment (zz-jason) [inline executor/builder.go]:
```suggestion
		return nil, errors.Errorf("inner child of Projection should be IndexLookupReader, but got %T", v)
```

File: executor/builder.go [hunk 2]
```diff
@@ -2121,6 +2123,35 @@ func (builder *dataReaderBuilder) buildIndexLookUpReaderForIndexJoin(ctx context
...
+	}
+	childExec, err := builder.buildIndexLookUpReaderForIndexJoin(ctx, physicalIndexLookUp, lookUpContents, indexRanges, keyOff2IdxOff, cwc)
+	if err != nil {
+		return nil, err
+	}
+
+	e := &ProjectionExec{
```
Comment (zz-jason) [inline executor/builder.go]: why not using `Projection.Init()`?

File: executor/builder.go [hunk 3]
```diff
@@ -2121,6 +2123,35 @@ func (builder *dataReaderBuilder) buildIndexLookUpReaderForIndexJoin(ctx context
...
+		baseExecutor:     newBaseExecutor(builder.ctx, v.Schema(), v.ExplainID(), childExec),
+		numWorkers:       builder.ctx.GetSessionVars().ProjectionConcurrency,
+		evaluatorSuit:    expression.NewEvaluatorSuite(v.Exprs, v.AvoidColumnEvaluator),
+		calculateNoDelay: v.CalculateNoDelay,
+	}
+
+	// If the calculation row count for this Projection operator is smaller
+	// than a Chunk size, we turn back to the un-parallel Projection
+	// implementation to reduce the goroutine overhead.
+	if int64(v.StatsCount()) < int64(builder.ctx.GetSessionVars().MaxChunkSize) {
```
Comment (zz-jason) [inline executor/builder.go]: The decision to use multiple workers is made in the execution phase in other places, it's better to unify them to the same layer.

Comment (lzmhhh123) [reply executor/builder.go]: This is in the execution phase.

File: executor/builder.go (hunk 2 above)
Comment (lzmhhh123) [reply executor/builder.go]: `ProjectionExec` doesn't have `init()`.

File: executor/builder.go [hunk 4]
```diff
@@ -2028,6 +2028,8 @@ func (builder *dataReaderBuilder) buildExecutorForIndexJoin(ctx context.Context,
 		return builder.buildIndexLookUpReaderForIndexJoin(ctx, v, lookUpContents, IndexRanges, keyOff2IdxOff, cwc)
 	case *plannercore.PhysicalUnionScan:
 		return builder.buildUnionScanForIndexJoin(ctx, v, lookUpContents, IndexRanges, keyOff2IdxOff, cwc)
+	case *plannercore.PhysicalProjection:
```
Comment (XuHuaiyu) [inline executor/builder.go]: We'd better add a comment here for when the inner child will be a PhysicalProjection

Comment (lzmhhh123) [reply executor/builder.go]: Good advice.

File: executor/builder.go [hunk 5]
```diff
@@ -2028,6 +2028,10 @@ func (builder *dataReaderBuilder) buildExecutorForIndexJoin(ctx context.Context,
 		return builder.buildIndexLookUpReaderForIndexJoin(ctx, v, lookUpContents, IndexRanges, keyOff2IdxOff, cwc)
 	case *plannercore.PhysicalUnionScan:
 		return builder.buildUnionScanForIndexJoin(ctx, v, lookUpContents, IndexRanges, keyOff2IdxOff, cwc)
+	// If the index lookup reader need to keep order. Then the planner will generate a projection as its parent.
```
Comment (XuHuaiyu) [inline executor/builder.go]: PkIsNotHandle is a requirement， but not mentioned here.

File: planner/core/find_best_task.go [hunk 6]
```diff
@@ -1023,6 +1023,8 @@ func (ds *DataSource) getOriginalPhysicalTableScan(prop *property.PhysicalProper
 	}.Init(ds.ctx, ds.blockOffset)
 	if ds.preferStoreType&preferTiFlash != 0 {
 		ts.StoreType = kv.TiFlash
+		ts.filterCondition = append(ts.filterCondition, ts.AccessCondition...)
```
Comment (XuHuaiyu) [inline planner/core/find_best_task.go]: Are these 2 lines necessary?

Comment (lzmhhh123) [reply planner/core/find_best_task.go]: Oh... That the code for another PR. I pushed the wrong branch.

File: executor/builder.go [hunk 7]
```diff
@@ -2028,6 +2028,10 @@ func (builder *dataReaderBuilder) buildExecutorForIndexJoin(ctx context.Context,
 		return builder.buildIndexLookUpReaderForIndexJoin(ctx, v, lookUpContents, IndexRanges, keyOff2IdxOff, cwc)
 	case *plannercore.PhysicalUnionScan:
 		return builder.buildUnionScanForIndexJoin(ctx, v, lookUpContents, IndexRanges, keyOff2IdxOff, cwc)
+	// If the index lookup reader need to keep order and the pk is not handle. Then the planner will generate a projection as its parent.
```
Comment (XuHuaiyu) [inline executor/builder.go]: Check will this be better?

The inner child of IndexJoin might be Projection when a combination of the following conditions is true:
1. The inner child fetch data using indexLookupReader
2. PK is not handle
3. The inner child needs to keep order

In this case, an extra column tidb_rowid will be appended in the output result of IndexLookupReader(see `copTask.doubleReadNeedProj`). Then we need a Projection upon IndexLookupReaderto prune the redundant column.

**Pull request merged**